protocol/keepalive/serverclient_test.go

```diff
@@ -0,0 +1,151 @@
+package keepalive_test
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	ouroboros "github.com/blinklabs-io/gouroboros"
+	"github.com/blinklabs-io/gouroboros/protocol"
+	"github.com/blinklabs-io/gouroboros/protocol/keepalive"
+	ouroboros_mock "github.com/blinklabs-io/ouroboros-mock"
+)
+
+const (
+	// MockKeepAliveWrongCookie is the wrong cookie for a keep-alive response
+	MockKeepAliveWrongCookie uint16 = 1000
+)
+
+// ConversationEntryKeepAliveResponse is a pre-defined conversation entry for a keep-alive response
+var ConversationEntryKeepAliveResponseWrongResponse = ouroboros_mock.ConversationEntryOutput{
+	ProtocolId: keepalive.ProtocolId,
+	IsResponse: true,
+	Messages: []protocol.Message{
+		keepalive.NewMsgKeepAliveResponse(MockKeepAliveWrongCookie),
+	},
+}
+
+// ConversationKeepAlive is a pre-defined conversation with a NtN handshake and repeated keep-alive requests
+// and responses
+var ConversationKeepAliveWrongResponse = []ouroboros_mock.ConversationEntry{
+	ouroboros_mock.ConversationEntryHandshakeRequestGeneric,
+	ouroboros_mock.ConversationEntryHandshakeNtNResponse,
+	ouroboros_mock.ConversationEntryKeepAliveRequest,
+	ConversationEntryKeepAliveResponseWrongResponse,
+}
+
+func TestServerKeepaliveHandling(t *testing.T) {
+	mockConn := ouroboros_mock.NewConnection(
+		ouroboros_mock.ProtocolRoleClient,
+		ouroboros_mock.ConversationKeepAlive,
+	).(*ouroboros_mock.Connection)
+
+	// Async mock connection error handler
+	asyncErrChan := make(chan error, 1)
+	go func() {
+		err := <-mockConn.ErrorChan()
+		if err != nil {
+			asyncErrChan <- fmt.Errorf("received unexpected error\n  got:   %v\n  wanted: no error", err)
+		}
+		close(asyncErrChan)
+	}()
+
+	oConn, err := ouroboros.New(
+		ouroboros.WithConnection(mockConn),
+		ouroboros.WithNetworkMagic(ouroboros_mock.MockNetworkMagic),
+		ouroboros.WithNodeToNode(true),
+		ouroboros.WithKeepAlive(true),
+		ouroboros.WithKeepAliveConfig(keepalive.NewConfig(
+			keepalive.WithKeepAliveFunc(func(protocolID uint16) error {
+				t.Log("keep alive beat")
+				return nil
+			}),
+			keepalive.WithPeriod(time.Millisecond*100),
+			// Set correct cookie
+			keepalive.WithCookie(ouroboros_mock.MockKeepAliveCookie),
+			keepalive.WithDoneFunc(func() error {
+				t.Log("keep alive done")
+				return nil
+			}),
+		)),
+	)
+	if err != nil {
+		t.Fatalf("unexpected error when creating Connection object: %s", err)
+	}
+
+	// Wait for mock connection shutdown
+	select {
+	case err, ok := <-asyncErrChan:
+		if ok {
+			t.Fatal(err.Error())
+		}
+	case <-time.After(2 * time.Second):
+		t.Fatalf("did not complete within timeout")
+	}
+
+	// Close connection
+	if err := oConn.Close(); err != nil {
+		t.Fatalf("unexpected error when closing Connection object: %s", err)
+	}
```

Comment:
There should be another wait for shutdown like the above with `oConn.ErrorChan()` instead of `asyncErrChan`